freedomofpress/securedrop-workstation
# Template consolidation via GUI updater

## Status
Ready for review

## Description of Changes

Closes #471. 

* Consolidations the SDW templates into two: `sd-small-buster-template` and `sd-large-buster-template`. 
* Handles migration via GUI updater
* Updates tests throughout

Depends on multiple closely related PRs:


* https://github.com/freedomofpress/securedrop-debian-packaging/pull/198
* https://github.com/freedomofpress/securedrop-proxy/pull/79
* https://github.com/freedomofpress/securedrop-export/pull/65
* https://github.com/freedomofpress/securedrop-client/pull/1160

## Template consolidation test plan

The procedure for upgrading prod machines will be as follows:

1. Proactively contact pilot participants to notify about special case update.
2. Release all packages, both RPMs (for dom0) and debs (for SDW VMs), at same time.
3. When GUI updater appears, users should close it, open a dom0 terminal, and run `sudo qubes-dom0-update -y`. Doing so will obtain the latest GUI updater code.
4. Users should then double-click the SecureDrop desktop icon to rerun the new GUI updater, which will handle the consolidation steps as part of the upgrade.

Given the special case of requesting user action, we'll need to test both the "happy" path, where users follow our advice exactly, and the "sad" path, where special action is not taken. 

See detailed test plan for the upcoming 0.5.0 release: https://github.com/freedomofpress/securedrop-workstation/wiki/0.5.0-Test-Plan Complete only the sections under **Scenario: Template consolidation.** The other scenarios we'll handle as part of QA, after merging this and related PRs.

## QA
Writing a separate "QA" checklist because merging this and its related PR requires careful orchestration. Prior to merge of this PR, there are test-only commits marked "TEMPORARY" that should be removed. Doing so is required to use the "main" channel of the apt-test repository. Right now, packages are only present in the "template-consolidation" channel to facilitate testing. 

The following PRs must be reviewed, then merged together:

* https://github.com/freedomofpress/securedrop-proxy/pull/79
* https://github.com/freedomofpress/securedrop-debian-packaging/pull/198
* PR moving the built packages to apt-test main channel

## Checklist

### If you have made code changes

      be left unchecked, as `flake8` also runs in CI)

### If you have made changes to the provisioning logic


      logic in `MANIFEST.in` and `rpm-build/SPECS/securedrop-workstation-dom0-config.spec`


### Switching a prod install to update from QA repositories
For testing purposes it may be necessary to modify a prod installation to use QA repos. The process of doing so is threefold:
- update /etc/apt/source.list.d/securedrop_workstation.list to use `apt-test.freedom.press` and add the testing key in all Debian-based SD templates
- update the repo definition in dom0 from `yum.securedrop.org` to `yum-test.securedrop.org`
- update /srv/salt/sd-default-config.yml to point the prod env definition at dev env variables.

To do so safely using Salt, as root:
-  untar `/home/user/securedrop-workstation/scripts/qa-switch.tar.gz` in /srv/salt
- run `bash /srv/salt/consolidation-qa-switch.sh`
